### PR TITLE
'replace' member name should not be treated as modifier

### LINKED ISSUE #11503: CS1585	Member modifier 'replace' must precede the member type and name
**Version Used**:

Bug seen in Roslyn Future  20160520.2 

**Steps to Reproduce**:
Copy and paste the below code

``` C#
using System.Text.RegularExpressions;

namespace TestCS1585
{
    public class Repro
    {
        public static string replace(string input, string regexp, string flags, string replacement)
        {

            RegexOptions options = RegexOptions.ECMAScript;

            if (flags.IndexOf("i") != -1)
            {
                options |= RegexOptions.IgnoreCase;
            }

            Regex regex = new Regex(regexp, options);

            if (flags.IndexOf("g") != -1)
            {
                return regex.Replace(input, replacement, 1);
            }
            else
            {
                return regex.Replace(input, replacement);
            }
        }
    }
}
```

**Expected Behavior**:

No errors

**Actual Behavior**:

Repro.cs(12,30,12,37): error CS1585: Member modifier 'replace' must precede the member type and name
Repro.cs(12,30,12,37): error CS1520: Method must have a return type

### PR DESCRIPTION
**Pull request by cston**
Fixes #11503

### CODE REVIEW

**Comment by cston** [comment]
@dotnet/roslyn-compiler, @mattgertz please review

File: src/Compilers/CSharp/Test/Syntax/Parsing/ReplaceParsingTests.cs [hunk 1]
```diff
@@ -100,5 +95,13 @@ private void OriginalInMember(string text, bool inReplace)
             var expectedKind = inReplace ? SyntaxKind.OriginalExpression : SyntaxKind.IdentifierName;
             Assert.Equal(expectedKind, expr.Kind());
         }
+
+        [Fact]
+        [WorkItem(11503, "https://github.com/dotnet/roslyn/issues/11503")]
+        public void MethodCalledReplace()
+        {
+            var root = SyntaxFactory.ParseCompilationUnit("class C { public static object replace() { } }", options: TestOptions.Regular.WithReplaceFeature());
+            root.Errors().Verify();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ReplaceParsingTests.cs]: Consider adding end to end test. Verify that binding works, the method can be called, etc.

**Comment by AlekseyTs** [comment]
LGTM

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 2]
```diff
@@ -2596,21 +2597,28 @@ private MemberDeclarationSyntax ParseMemberDeclarationOrStatement(SyntaxKind par
...
+                type = (explicitInterfaceOpt == null)
+                    ? (TypeSyntax)newType
+                    : _syntaxFactory.QualifiedName(explicitInterfaceOpt.Name, explicitInterfaceOpt.DotToken, newType);
+                explicitInterfaceOpt = null;
+                identifierOrThisOpt = default(SyntaxToken);
+                typeParameterListOpt = default(TypeParameterListSyntax);
+                return true;
+            }
+
+            return false;
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: Not a blocker. Consider using the negative condition and returning false right then and there.

File: src/Compilers/CSharp/Test/Syntax/Parsing/MoreReplaceParsingTests.cs [hunk 3]
```diff
@@ -0,0 +1,766 @@
...
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Roslyn.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.CSharp.UnitTests
+{
+    // Adapted from AsyncParsingTests.
+    [CompilerTrait(CompilerFeature.SourceGenerators)]
+    public class MoreReplaceParsingTests : ParsingTests
+    {
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Syntax/Parsing/MoreReplaceParsingTests.cs]: Empty line below.

**Comment by jcouv** [comment]
:shipit:

**Comment by jcouv** [comment]
It seems many of the unusual but valid parsing cases ("replace replace replace()") could also be turned into CompileAndVerify tests.

File: src/Compilers/CSharp/Test/Syntax/Parsing/ReplaceParsingTests.cs (hunk 1 above)
Comment (cston) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ReplaceParsingTests.cs]: Good suggestion, thanks. I'll add in a separate PR.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 2 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: Done.

File: src/Compilers/CSharp/Test/Syntax/Parsing/MoreReplaceParsingTests.cs (hunk 3 above)
Comment (cston) [inline src/Compilers/CSharp/Test/Syntax/Parsing/MoreReplaceParsingTests.cs]: Removed.

File: src/Compilers/CSharp/Test/Syntax/Parsing/ReplaceParsingTests.cs (hunk 1 above)
Comment (cston) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ReplaceParsingTests.cs]: Added end to end test.

**Comment by MattGertz** [comment]
Where are the test runs?

**Comment by jaredpar** [comment]
@MattGertz click on the "view details" button.

**Comment by MattGertz** [comment]
Yes, I got that, but I'm seeing "0 of 4 checks passed."

**Comment by MattGertz** [comment]
I guess they must have passed?  I have to drill in on each one to see that...

**Comment by jaredpar** [comment]
@MattGertz they did pass.  I don't know why Github is showing them in the running state.

**Comment by VSadov** [comment]
LGTM

**Pull request merged**